text/0000-let-expression.md

```diff
@@ -0,0 +1,626 @@
+// generalized matches! macro:
+assert!((let Some(x) = y) && x > 2);
+
+// generalized let-else construct
+(let Some(a) = b) && (let Some(c) = f(&a)) || {
+    return Err("failed");
+};
+println!("{}, {}", a, c);
+
+// generalized assignment with default
+(let Some(x) = y) || (let Foo(x) = bar) || (let x = default);
+println!("{}", x);
+```
+
+# Motivation
+[motivation]: #motivation
+
+The main motivation for this RFC is improving consistency and ergonomics of language.
+
+## Consistency
+
+Currently we have `if`, `if let` and `if let && let` and we teach them as three different
+constructs (plus `let else` in future). But if we make `let` a `bool` expression, all will become the same and it would be
+easier for new learners to get it. After this RFC you get only an unused parenthesis warning
+for `if (let Some(x) = y) { }`, not a hard error. And it will have the same behavior with
+classic `if let`. This is actually [a mistake by a new learner](https://github.com/rust-lang/rust/issues/82827)
+that show us new learners expect it.
+
+This situation is worse with `if let chain` that mix let expressions with `&&` and
+other bools. In fact the compiler will understand it via interpreting let as an
+expression, so why we force humans to understand it another way?
+
+This proposal is also in-line with "everything is an expression" that we have
+in rust.
+
+## Ergonomics
+
+It also available many super powers for us that can
+help decreasing rightward drift without adding to implementation and understanding complexity, and
+actually decreasing it by removing `let-else` and preventing from future similar constructs.
+
+### Compare to `let-else`
+
+```rust
+// simple let else
+let Some(x) = y else {
+    return Err("fail");
+};
+
+// with let expression
+(let Some(x) = y) || {
+    return Err("fail");
+};
+
+// or even better
+(let Some(x) = y) || return Err("fail");
+
+// let else else future possiblity
+let Some(x) = a else b else c else { return; };
+
+// with let expression
+(let Some(x) = a) || (let Some(x) = b) || (let Foo(x) = bar) || { return; };
+```
+
+### New constructs
+
+```rust
+// reuse if let body
+if let Some(x) = a || let Some(x) = b || Ok(x) = c {
+    // body with many lines of code
+} else {
+    // else with many lines of code
+}
+
+// today alternative without code duplication:
+if let ((Some(x), _, _) | (_, Some(x), _) | (_, _, Ok(x)) = (a, b, c) {
+
+// assignment with default
+(let Foo(x) = a) || (let Bar(x) = b) || (let x = default);
+
+// today alternative:
+let x = if let Foo(x) = a { x } else if let Bar(x) = b { x } else { default };
+
+// simple let expression
+assert!((let Some(x) = a) && (let Some(y) = b(x)) && x == y);
+
+// today alternative with if let chains:
+assert!(matches!(a, Some(x) if let Some(y) = b(x) && x == y));
+assert!(if let Some(x) = a && let Some(y) = b(x) && x == y { true } else { false });
+```
+
+## Why now?
+This RFC exists thanks to people who choose `if let` for syntax we know today. If those
```

Comment:
This section feels way too aggressive towards the people that originally introduced the if let syntax. The syntax wasn't picked randomly and there's an RFC that explains it https://rust-lang.github.io/rfcs/0160-if-let.html . It was inspired by a similar syntax in swift. It's fine to not like the decision, but it's certainly not a random decision. 

I'd also suggest using a less graphic analogy to say that this RFC would not have been necessary.

Reply:
This RFC is pro `if let` and I like this syntax. Random isn't used as a negatively charged word, and from it I meant there were other valid choices, but they choose `if let` for some good reasons other than a future let expression RFC, so relatively random, to this RFC. If the text does not convey this, help me change it and use a better word instead of random.

Reply:
I tried to address your comment in [this commit](https://github.com/rust-lang/rfcs/pull/3159/commits/c2bdecdf938151c27abc0f8cc396141215963fba).

Reply:
Here's what I would suggest. I still feel that the tone is a bit too opinionated, especially the last sentence, but I tried to keep it anyway.

```
This RFC exists because of the `if let` syntax we know today.
That syntax wasn't the only choice and there was other options like `iflet`, `if match`, `let if`, `if is` or another `keyword`. 
If one of those had been picked this RFC would not have been necessary or would have been very different.
If the `&&` or `,` symbol is used for the `if let chain` RFC, this RFC would not be necessary since it would not be compatible.

But luck is not always with us. We can't expect each new RFC to add another piece of the let expression puzzle to the language. For example, `matches!` and `let-else` are potentially not compatible with this RFC, `matches!` can [coexist with let expression][future-of-matches] but let-else is not compatible with let expression and therefore can't coexist. Some people have felt that `let-else` is not compatible with `if let chain`, this is one of the unresolved questions. The answer to this question is: they are not compatible! This RFC has less additions to the language grammar and is more expressive.

The goal of proposing this now is to prevent `let-else` and future similar RFCs to be stabilized. Originally, authors of the `if let chain` RFC had an incremental plan toward let expression.
The implementation of the first part took a long time and is still not done. Since the `if let chain` RFC is still not completed we see RFCs like `let-else` that tries to fix the missing parts of the `if let chain` RFC.

Even if it doesn't fit in this year road-map, we should decide if we want it or not today. Even today
is too late!
```

Reply:
@HKalbasi not sure if you've seen this since it was posted in a resolved section

Reply:
@IceSentry I didn't see, and even when you mentioned me I was seeing a notification from nowhere :)

Thanks for suggestion, I updated this section with your text as base. Is it now good?